src/services/codefixes/importFixes.ts

```diff
@@ -437,35 +571,38 @@ namespace ts.codefix {
         program: Program,
         sourceFile: SourceFile,
         position: number | undefined,
-        preferTypeOnlyImport: boolean,
+        isValidTypeOnlyUseSite: boolean,
         useRequire: boolean,
         moduleSymbols: readonly SymbolExportInfo[],
         host: LanguageServiceHost,
         preferences: UserPreferences,
         fromCacheOnly?: boolean,
-    ): { computedWithoutCacheCount: number, fixes: readonly (FixAddNewImport | FixUseImportType)[] } {
+    ): { computedWithoutCacheCount: number, fixes: readonly (FixAddNewImport | FixAddJsdocTypeImport)[] } {
         const isJs = isSourceFileJS(sourceFile);
         const compilerOptions = program.getCompilerOptions();
         const moduleSpecifierResolutionHost = createModuleSpecifierResolutionHost(program, host);
-        const checker = program.getTypeChecker();
+        const getChecker = memoizeOne((isFromPackageJson: boolean) => isFromPackageJson ? host.getPackageJsonAutoImportProvider!()!.getTypeChecker() : program.getTypeChecker());

```

Comment:
> `!()!`

-nervous laugh-

Reply:
This is an invariant used all over this file—`isFromPackageJson` literally means the symbol came from the checker we’re asserting the existence of with the `!`’s.